toplevel/topcommon.ml

```diff
@@ -380,3 +380,41 @@ let try_run_directive ppf dir_name pdir_arg =
             dir_name dir_type arg_type;
           false
   end
+
+(* Overriding exception printers with toplevel-specific ones *)
+
+let loading_hint_printer ppf err =
+  Symtable.report_error ppf err;
+  match err with
+  | Symtable.Undefined_global s ->
+    if !Sys.interactive then begin
+      let find_with_ext ext =
+        try Some (Load_path.find_uncap (s ^ ext)) with Not_found -> None
+      in
+      fprintf ppf
+        "@.Hint: @[\
+         This means a module's interface is loaded, \
+         but its implementation is not.@,";
+      begin match List.find_map find_with_ext [".cma"; ".cmo"] with
+      | Some path ->
+        fprintf ppf
+          "Found %s @,in the load paths. \
+           @,Did you mean to load it using @,#load \"%s\" \
+           @,or by passing it as an argument to the toplevel?"
+           path (Filename.basename path)
+      | None ->
+        fprintf ppf
+          "Did you mean to load a compiled implementation of the module \
+           @,using #load or by passing it as an argument to the toplevel?"
+      end;
+      fprintf ppf "@]"
+    end
+  | _ -> ()
```

Comment:
This printer, given the way it is registered, claims to support all `Symtable.Error` cases, but in fact ignores all errors that are not `Undefined_global`. This is awkward and also probably wrong. I would feel safer if the `loading_hint_printer` was passed directly to `register_error_of_exn`, and would return `Some` only in the case that it definitely knows how to handle better than the standard printers.